### PR TITLE
[BOLT] Fix runOnEachFunctionWithUniqueAllocId

### PR DESCRIPTION
**Pull request by kbeyls**
When runOnEachFunctionWithUniqueAllocId is invoked with ForceSequential=true, then the current implementation runs the function with AllocId==0, which is the Id for the shared, non-unique, default AnnotationAllocator.

However, the documentation for runOnEachFunctionWithUniqueAllocId states:
```
/// Perform the work on each BinaryFunction except those that are rejected
/// by SkipPredicate, and create a unique annotation allocator for each
/// task. This should be used whenever the work function creates annotations to
/// allow thread-safe annotation creation.
```

Therefore, even when ForceSequential==true, a unique AllocId should be used, i.e. different from 0.

In the current upstream BOLT this is presumably not depended on, but it is needed to reduce memory usage for analyses that use a lot of memory/annotations. Examples are the pac-ret and stack-clash analyses that currently have prototype implementations as described in https://discourse.llvm.org/t/rfc-bolt-based-binary-analysis-tool-to-verify-correctness-of-security-hardening/78148 These analyses use the DataFlowAnalysis framework to sometimes store quite a lot of information on each MCInst. They run in parallel on each function. When the dataflow analysis is finished, the annotations on each MCInst can be removed, hugely saving on memory consumption. The only annotations that need to remain are those that indicate some unexpected properties somewhere in the binary.

Fixing this bug enables implementing the deletion of the memory used by those huge number of DataFlowAnalysis annotations (by invoking BC.MIB->freeValuesAllocator(AllocatorId)), even when run with --no-threads. Without this bug fixed, the invocation of BC.MIB->freeValuesAllocator(AllocatorId) results in also the memory for all other annotations to be deleted, as AllocatorId is 0.

### CODE REVIEW

**Comment by kbeyls** [comment]
The one precommit test failure on `buildkite/github-pull-requests` looks exactly the same as on this other recent BOLT-related PR #90004  (RISCV/fake-label-no-entry.c). Therefore, It seems that the precommit test failure is not related to the patch in this PR.

File: bolt/lib/Core/ParallelUtilities.cpp [hunk 1]
```diff
@@ -164,6 +164,15 @@ void runOnEachFunction(BinaryContext &BC, SchedulingPolicy SchedPolicy,
   Pool.wait();
 }
 
+static void EnsureAllocatorExists(BinaryContext &BC, unsigned AllocId) {
```
Comment (maksfb) [inline bolt/lib/Core/ParallelUtilities.cpp:167]:
```suggestion
static void ensureAllocatorExists(BinaryContext &BC, unsigned AllocId) {
```

Comment (maksfb) [reply bolt/lib/Core/ParallelUtilities.cpp:167]: Better make it a closure in `runOnEachFunctionWithUniqueAllocId()` unless you plan to use it more in follow-up patches.

File: bolt/lib/Core/ParallelUtilities.cpp [hunk 2]
```diff
@@ -164,6 +164,15 @@ void runOnEachFunction(BinaryContext &BC, SchedulingPolicy SchedPolicy,
...
+  if (!BC.MIB->checkAllocatorExists(AllocId)) {
+    MCPlusBuilder::AllocatorIdTy Id =
+        BC.MIB->initializeNewAnnotationAllocator();
+    (void)Id;
+    assert(AllocId == Id && "unexpected allocator id created");
```
Comment (maksfb) [inline bolt/lib/Core/ParallelUtilities.cpp:172]: nit:
```suggestion
    assert(AllocId == Id && "Unexpected allocator id created");
```

**Review by maksfb — COMMENTED** [review]
Thanks for the fix. Looks good overall.

File: bolt/lib/Core/ParallelUtilities.cpp (hunk 1 above)
Comment (kbeyls) [reply bolt/lib/Core/ParallelUtilities.cpp]: Ah, yes, it's better as a closure.
I've now moved it to a closure, see the latest commit.

**Review by maksfb — APPROVED** [review]

**Pull request merged**